applications/nrf_desktop/src/modules/dvfs.c

```diff
@@ -0,0 +1,270 @@
+{
+	int32_t ret = dvfs_service_handler_change_freq_setting(target_freq);
+
+	if (ret) {
+		handle_dvfs_error(ret);
+	} else {
+		current_freq = target_freq;
+		LOG_INF("Have requested %s frequency",
+			get_dvfs_frequency_setting_name(target_freq));
+		cancel_dvfs_retry_work();
+	}
+}
+
+static enum dvfs_frequency_setting check_required_frequency(void)
+{
+	if (dfvs_requests_state & dvfs_high_freq_bitmask) {
+		return DVFS_FREQ_HIGH;
+	} else if (dfvs_requests_state & dvfs_medlow_freq_bitmask) {
+		return DVFS_FREQ_MEDLOW;
+	} else {
+		return DVFS_FREQ_LOW;
+	}
+}
+
+static void process_dvfs_states(enum dvfs_state state, bool turn_on)
+{
+	if (module_state == STATE_ERROR) {
+		return;
+	}
+
+	if (turn_on) {
+		dfvs_requests_state |= BIT(state);
+		LOG_DBG("%s ACTIVE", get_dvfs_state_name(state));
+	} else {
+		dfvs_requests_state &= ~BIT(state);
+		LOG_DBG("%s NOT ACTIVE", get_dvfs_state_name(state));
+	}
+
+	enum dvfs_frequency_setting required_freq = check_required_frequency();
+
+	if ((required_freq != current_freq) &&
+	    (!k_work_delayable_is_pending(&dvfs_retry_data.dvfs_retry_work))) {
+		set_dvfs_freq(required_freq);
+	} else if ((required_freq == current_freq) &&
+	    (k_work_delayable_is_pending(&dvfs_retry_data.dvfs_retry_work))) {
+		cancel_dvfs_retry_work();
+	}
+}
+
+static bool handle_ble_peer_conn_params_event(const struct ble_peer_conn_params_event *event)
+{
+	if (!event->updated) {
+		/* Ignore the connection parameters update request. */
+		return false;
+	}
+
+	__ASSERT_NO_MSG(event->interval_min == event->interval_max);
+
+	process_dvfs_states(DVFS_STATE_LLPM_CONNECTED,
+			    event->interval_min & REG_CONN_INTERVAL_LLPM_MASK);
+
+	return true;
+}
+
+static void dvfs_retry_work_handler(struct k_work *work)
+{
+	LOG_DBG("Retrying to change DVFS frequency.");
+
+	enum dvfs_frequency_setting required_freq = check_required_frequency();
+
+	__ASSERT_NO_MSG(required_freq != current_freq);
+
+	set_dvfs_freq(required_freq);
+}
+
+static void get_req_modules(struct module_flags *mf)
+{
+	module_flags_set_bit(mf, MODULE_IDX(main));
+#if CONFIG_CAF_SETTINGS_LOADER
+	module_flags_set_bit(mf, MODULE_IDX(settings_loader));
+#endif
+#if CONFIG_DESKTOP_BLE_BOND_ENABLE
+	module_flags_set_bit(mf, MODULE_IDX(ble_bond));
+#endif
+}
+
+static bool app_event_handler(const struct app_event_header *aeh)
+{
+	static struct module_flags req_modules_bm;
+
+	if (is_module_state_event(aeh)) {
+		const struct module_state_event *event = cast_module_state_event(aeh);
+
+		if (check_state(event, MODULE_ID(main), MODULE_STATE_READY)) {
+			__ASSERT_NO_MSG((dvfs_high_freq_bitmask & dvfs_medlow_freq_bitmask) == 0);
+			k_work_init_delayable(&dvfs_retry_data.dvfs_retry_work,
+					      dvfs_retry_work_handler);
+			get_req_modules(&req_modules_bm);
+
+			/* In case user implemented empty get_req_modules function */
+			if (module_flags_check_zero(&req_modules_bm)) {
+				process_dvfs_states(DVFS_STATE_INITIALIZING, false);
+				state = STATE_READY;
```

Comment:
The `state = STATE_READY`; assignment should be done regardless of the `if (module_flags_check_zero(&req_modules_bm)) {` condition (otherwise you still actually do not set `state = STATE_READY;` ...)

Reply:
I thought we enter state ready when we finish waiting for other modules to initialize